flang/include/flang/Optimizer/Dialect/FIROps.td

```diff
@@ -134,7 +134,8 @@ def fir_AllocaOp : fir_Op<"alloca", [AttrSizedOperandSegments,
     OptionalAttr<StrAttr>:$bindc_name,
     UnitAttr:$pinned,
     Variadic<AnyIntegerType>:$typeparams,
-    Variadic<AnyIntegerType>:$shape
+    Variadic<AnyIntegerType>:$shape,
+    OptionalAttr<UnitAttr>:$omp_threadprivate_decl
```

Comment:
I'm not sure we want to add this kind of attribute on the FIR operation. If really needed it should be an opaque attribute only known by the OpenMP dialect.

Reply:
Agreed. The FIR Op definition is not the right place to put OMP attributes. OMP lowering can add attributes to Ops.

I also have a semantics comment. `fir.alloca` is defined to mean allocating space on the stack, and the stack is already "thread local" in traditional threading semantics. So this attribute really has no effect on a stack allocated variable. But I think what is really desired here is _static thread local_ variable allocation where the memory allocation will be done in the thread's local storage area. That semantics is not defined in FIR. (FIR has stack, heap, and shared global allocation.) So, it seems like we really need a new op to be defined for TLS globals/statics.

Reply:
I am not sure what you mean exactly @PeixinQiao .

We could add a `<dialect>.tls_variable` under whatever dialect seems most appropriate. (That might be FIR, OpenMP, or something else.)

It would be similar in structure to a `fir.global`, but have distinct semantics.

It seems like it makes more sense to make it a distinct op type because thread-local storage and global storage do not have the same semantics. Making it a distinct op makes it (1) obvious, (2) something optimization writers have to consider and think through, and (3) doesn't change the semantics of already defined Ops to be something else entirely under special conditions.

Reply:
There are two methods to implement the OpenMP threadprivate directive. One is to use the runtime call (`__kmpc_threadprivate_cached`) to get the value of the corresponding thread, and the other one is to use TLS. For clang, the default usage is TLS, and the option `-fnoopenmp-use-tls` is needed to use the runtime call. For classic-flang, the default usage is the runtime call, and the TLS seems not implemented yet.

For the following c code:
```
double counter = 0;
#pragma omp threadprivate(counter)
double increment_counter() {
  counter = counter * 2;
  return counter;
}
```
Here are the IR looks like generated by two methods:
```
$ clang -fopenmp -x c++ -emit-llvm -S -fexceptions -fcxx-exceptions example.c -o example.ll && cat example.ll

$_ZTW7counter = comdat any
@counter = dso_local thread_local global double 0.000000e+00, align 8
...
  %0 = load double, double* @counter, align 8
  %mul = fmul contract double %0, 2.000000e+00
```
```
$ clang -fopenmp -x c++ -emit-llvm -S -fexceptions -fcxx-exceptions example.c -fnoopenmp-use-tls -o example.ll && cat example.ll

@counter = dso_local global double 0.000000e+00, align 8
@counter.cache. = common global i8** null
...
  %0 = call i32 @__kmpc_global_thread_num(%struct.ident_t* @1)
  %1 = call i8* @__kmpc_threadprivate_cached(%struct.ident_t* @1, i32 %0, i8* bitcast (double* @counter to i8*), i64 8, i8*** @counter.cache.)
  %2 = bitcast i8* %1 to double*
  %3 = load double, double* %2, align 8
  %mul = fmul contract double %3, 2.000000e+00
```
For the following fortran code:
```
program main
  integer, save :: x
  !$omp threadprivate(x)
  x = 1
end
```
Here are the IR looks like generated by two methods:
```
$ flang -fopenmp test.f90 -S -emit-llvm && cat test.f90
%struct.BSS1 = type <{ [16 x i8] }>
...
  %3 = bitcast %struct.BSS1* @.BSS1 to i8*, !dbg !8
  %4 = getelementptr i8, i8* %3, i64 8, !dbg !8
  %5 = bitcast i8* %4 to i64*, !dbg !8
  %6 = call i8* @__kmpc_threadprivate_cached(i64* null, i32 %1, i64* %2, i64 4, i64* %5), !dbg !8
```
```
$ flang -fopenmp test.f90 -S -emit-llvm -Mx,69,0x80 && cat test.f90
%struct.BSS1 = type <{ [4 x i8] }>
@TPpx_ = common thread_local global i8* null, align 64
  %3 = bitcast i8** @TPpx_ to i8*, !dbg !8
  %4 = call i8* @_mp_get_threadprivate(i8* null, i32 %1, i8* %2, i64 4, i8* %3), !dbg !8
```
I do not find `_mp_get_threadprivate` call in llvm-project.

I start with the runtime call. I have created one new Op in OpenMP Dialect and finished lowering from parse-tree to MLIR. It should work, and I will update this patch soon. I think this implementation does not need something like `<dialect>.tls_variable`, but let's see when I finish the whole lowering flow and update it.

For the implementation using TLS technique such as clang usage (https://github.com/llvm/llvm-project/blob/2542bfa43a9709982ee20f8d86f905f4115c41f9/clang/lib/CodeGen/CGOpenMPRuntime.cpp#L2021-L2025), we may need one Op like `<dialect>.tls_variable` @schweitzpgi mentioned, which should have the similar structure as `fir.global` as the above clang IR showed.

BTW, once two methods are both supported, the performance should be tested and set the better one as deault while giving the other one the option.